File: Tactics-iOS/Classes/Level/Terrain/Terrain.m
//
//  Terrain.m
//  Tactics-iOS
//
//  Created by Amos, Grant ext (E IT S MCC) on 6/22/14.
//  Copyright (c) 2014 SomethingCool. All rights reserved.
//

#import "Terrain.h"
#import "HexagonFactory.h"

#import "CC3Mesh.h"
#import "CC3Material.h"
#import "CC3Texture.h"
#import "CC3Node.h"

@interface Terrain() {
    int _width;
    int _height;
    int **_map;
}

@end

@implementation Terrain

- (id)initWithWidth:(int)width height:(int)height andMap:(int**)map
{
    self = [super init];
    
    if (self)
    {
        _width = width;
        _height = height;
        _map = map;
        
        [self setup];
    }
    
    return self;
}

- (void)dealloc
{
    free(_map);
}

- (void)setup
{
    float height = 0.5;
    float radius = 1.0;
    float hexHeight = 2 * radius;
    float rowHeight = 1.5 * radius;
    float halfWidth = sqrt((radius * radius) - (radius / 2) * (radius / 2));
    float hexWidth = 2 * halfWidth;
    float extraWidth = hexHeight - rowHeight;
    
    CC3Texture *texture = [CC3Texture textureFromFile:@"Hex256.png"];
    
    for (int x = 0; x < _width; x++)
    {
        for (int y = 0; y < _height; y++)
        {
            CC3Node *hexNode = [HexagonFactory.factory
                                createNodeWidthRadius:radius
                                height:height];
            
            float xPos = x * hexWidth + ((y & 1) * halfWidth);
            float yPos = y * rowHeight;
            
            hexNode.location = cc3v(xPos, 0, yPos);
            //hexNode.texture = texture;
            hexNode.diffuseColor = ccc4f(0.8, 0.8, 0.8, 1.0);
            
            [self addChild:hexNode];
        }
    }
}

@end

File: Tactics-iOS/Classes/Level/Terrain/Terrain.swift
//
//  Terrain.swift
//  tactics-ios
//
//  Created by Amos, Grant ext (E IT S MCC) on 6/19/14.
//  Copyright (c) 2014 SomethingCool. All rights reserved.
//

import Foundation
import SceneKit

class Terrain: SCNNode {
    
    var width: Int
    var height: Int
    var map: Int[][]
    
    init(width: Int, height: Int, map: Int[][]) {
        self.width = width
        self.height = height
        self.map = map
        
        super.init()
        
        var radius = 0.5
        var hexHeight = 2 * radius
        var rowHeight = 1.5 * radius
        var halfWidth = sqrt((radius * radius) - ((radius / 2) * (radius / 2)))
        var hexWidth = 2 * halfWidth
        var extraHeight = hexHeight - rowHeight
        
        //create boxes for hexagons temporarily
        let hexGeometry1 = HexagonGeometry.CreateWithRadius(Float(radius), andHeight: 0.5)
        let hexGeometry2 = HexagonGeometry.CreateWithRadius(Float(radius), andHeight: 1)
        let hexGeometry3 = HexagonGeometry.CreateWithRadius(Float(radius), andHeight: 1.5)
        // create and configure a material
        let material = SCNMaterial()
        material.diffuse.contents = UIImage(named: "Hex256")
        material.locksAmbientWithDiffuse = true
        
        for (x, column) in enumerate(map) {
            for (y, value) in enumerate(column) {
                var hexNode = SCNNode()
                
                hexNode.position.y = 0.0
                hexNode.position.x = Float(x) * Float(hexWidth) + Float((Double(y & 1) * halfWidth))
                hexNode.position.z = Float(y) * Float(rowHeight)
                
                if value == 0 {
                    hexNode.geometry = hexGeometry1
                } else if value == 1 {
                    hexNode.geometry = hexGeometry2
                } else if value == 2 {
                    hexNode.geometry = hexGeometry3
                }
                
                // set the material to the 3d object geometry
                hexNode.geometry.firstMaterial = material
                
                //Add it to the terrain's child nodes
                self.addChildNode(hexNode)
            }
        }
    }
}